### ISSUE TITLE
The Change Management - Approval Process

### ISSUE THREAD

**Issue by kyrylo-hulak** [aws-ia/terraform-aws-control_tower_account_factory#153]
**Describe the outcome you'd like**
It would be great to have there "Approval Process" set before "Applying Changes" with terraform tool.
Once there's got executed a terraform plan to be able to review it. Like adding one more stage in the pipeline to send a notification via SNS for the approval process. Since, to provision a new account is the base for other deployments and it's crucial to ensure that terraform plan looks like you were expected. At the moment, All changes has an "--auto-approve" flow. It might be fine for the sandbox environments, another is for prod accounts.
Thank you!

**Comment by snebhu3** [comment]
@kyrylo-hulak thank you for the feature request, use-case and context.
I have recorded your request in our internal backlog.

**Comment by kyrylo-hulak** [comment]
thank you, @snebhu3!

**Comment by kyrylo-hulak** [comment]
Please use an attached file to have visualized the enhancement
![The Change Management - Aproval Process](https://user-images.githubusercontent.com/88105512/168619663-af1fc582-efe3-4dbd-b4fe-b463b7d97e77.png)

**Comment by v-rosa** [comment]
@kyrylo-hulak this request is not only useful for the `account-request` flow, it is also essential for the customization pipelines.

@snebhu3 Should I create a different issue to address the approval process for customization pipelines?

**Comment by snebhu3** [comment]
@v-rosa thank you for reaching out. I have recorded your request in the existing backlog.

**Comment by MichaelYak** [comment]
@snebhu3 Is there a timeline for this feature? customizations pipeline is not of much use in production scenario without ability to review terraform plan prior to apply. PS: There is no easy workaround that'd allow us to add this feature and later to easily migrate to the upstream repo when this feature is added. It requires modifying aft-create-pipeline buildspec + fetching customizations_pipeline code from a private repository.

**Comment by v-rosa** [comment]
@MichaelYak as a **dirty workaround** you can modify the buildspecs to add a _long enough_ sleep between tf plan and tf apply. Customizations pipeline triggers are manually by design so this reduces the risk of blind terraform apply. Again this is a **dirty workaround**.

**Comment by kyrylo-hulak** [comment]
In that case, it would be great to have the process in place, out-of-box. Without a need to make tweaks and tunes, the integrity of the solution would be impacted. Also, having the feature in place oversees a company's compliance and controls the overall deployment. Nowadays, without control through a breach, we can get a number of not expected resources running. Without the process, we blindly accept all changes coming from a git repo without approval!

**Comment by v-rosa** [comment]
> Without the process, we blindly accept all changes coming from a git repo without approval!

This is partially true. I'll try to explain my PoV.

- The customization pipeline only runs with a manual trigger on the AFT account, this is by desing.
  - access to this account it's (or should be) already very restricted.
- An approval process on those customization repositories should be enforced, e.g. mandatory multi PR reviewers when merging into master.

Yet, I totally agree this missing security feature is a show stopper to use in production systems.

**Comment by MichaelYak** [comment]
@v-rosa I sort of agree with your PoV that it should not be needed to make a review of ter. plan after the code is merged into master. That means that the review needs to be done before the actual merge, which in turn means that ter. plan needs to be run on the change during/before the review process. (in my opinion it is essential to be able see the output for ter. plan in order to be able to make a judgment on whether to approve PL or not). It gives two problems: 1) to build up environment for ter. plan to run 2)there might occur race conditions if ter. plan is to be run on a feature branch, while apply - master (from aft codebuild context).

**Comment by v-rosa** [comment]
> this request is not only useful for the account-request flow, it is also essential for the customization pipelines.
> @v-rosa thank you for reaching out. I have recorded your request in the existing backlog.

@snebhu3 Do you have any news about this request?

**Comment by alibvr** [comment]
Hi, any update on this? or another alternative to enable terraform plan locally?

**Comment by azbpa** [comment]
@alibvr you can mimic what the [pipeline](https://github.com/aws-ia/terraform-aws-control_tower_account_factory/blob/main/modules/aft-customizations/buildspecs/aft-account-customizations-terraform.yml) is doing locally, for example inside a VSCode Dev Container. In combination with tools like direnv you can set different context for each customization folder and/or member account. Thereby you can render jinja2 templates and set required environment variables dynamically. If you read the account number from the stdin every time you switch into a customization folder, you can easily run any terraform command against any account. Of course you will need a privileged user that is permitted to assume the account individual AFTExecution role through the AFTAdmin role in AFT management account.

**Comment by Drewster727** [comment]
Hi - we just started using AFT a few weeks ago and realized how much of a significant issue this is. We are dividing up account customizations by business unit + environment. The fact that there is no ability to easily plan changes locally and more importantly, execute a terraform plan in a pull request makes AFT a tough sell on usage. Any ideas?

**Comment by v-rosa** [comment]
Welcome to the long waiting list :) 

You may try to implement on your own, the author of [this issue](https://github.com/designravenous) shared his approach to the problem.

**Comment by Drewster727** [comment]
Hi @v-rosa your issue link just takes me to a user profile. Can you resend the correct link? Thank you!

**Comment by seansilvestri** [comment]
@alibvr did you create an actual module for this? I'm looking at trying to get the script approach to run but for the aft-account-customizations repo, seems I would have to do this for every customization directory, which is a bit tedious. The example above references the aft-global-customizations which only has one terraform directory

**Comment by petrosmelachrinos** [comment]
> Hi, any update on this? or another alternative to enable terraform plan locally?

Any updates on this feature?

In our case what we did for testing before shipping was to simulate the aft process.
We have created a new AWS account let's say `aft-testing`

We have created 2 layers, **Account Request Layer** and **Account Customizations Layer**.

## Account Request Layer
1) In `aft-testing/account-request/aft-ssm.tf`:
```hcl
resource "aws_ssm_parameter" "networking_config" {
  name = "/aft/account-request/custom-fields/networking-config"
  type = "String"
  insecure_value = jsonencode({
    public_subnets          = var.at_public_subnets
    private_subnets         = var.at_private_subnets
    intra_subnets           = var.at_intra_subnets
    transit_gateway_subnets = var.at_transit_gateway_subnets
  })
}
```
2) Configuration in `aft-testing/account-request/aft-testing-account.auto.tfvars`:
```hcl
at_public_subnets = [
  {
    cidr_block        = "10.x.x.x"
    availability_zone = "eu-central-1a"
  }
]
```
3) Define vars in `aft-testing/account-request/aft-testing-account-variables.tf` 

The above simulates the AFT account request process in which the ssm parameters are equivalent with AFT's `custom_fields`.

In AFT account request repo the associate config would be similar to the below:
```hcl
custom_fields = {
    networking-config = jsonencode({
      public_subnets          = var.sct_public_subnets
      private_subnets         = var.sct_private_subnets
      intra_subnets           = var.sct_intra_subnets
      transit_gateway_subnets = var.sct_transit_gateway_subnets
    })
}
```
aft-testing-account.auto.tfvars and `aft-testing/account-request/aft-testing-account-variables.tf` can remain the same.

## Account Customizations Layer
In `aft-testing/account-customization/main.tf`:
```hcl
data "aws_ssm_parameter" "networking_config" {
  name = "/aft/account-request/custom-fields/networking-config"
}

locals {
  networking_config       = jsondecode(nonsensitive(data.aws_ssm_parameter.networking_config.value))
  public_subnets          = local.networking_config.public_subnets
  private_subnets         = local.networking_config.private_subnets
  intra_subnets           = local.networking_config.intra_subnets
  transit_gateway_subnets = local.networking_config.transit_gateway_subnets
}

module "vpc-workload-accounts" {
  source = "./modules/vpc-workload-accounts"
  public_subnets          = local.public_subnets
  private_subnets         = local.private_subnets
  intra_subnets           = local.intra_subnets
  transit_gateway_subnets = local.transit_gateway_subnets
}
```
For account customizations, same code can be used in in the  AFT account customization repo.

**Comment by v-rosa** [comment]
@kyrylo-hulak not sure if you're still using AFT, but in case you still need this feature take a look here: https://github.com/aws-ia/terraform-aws-control_tower_account_factory/issues/636
